packages/tooling/src/lib/flatten-schema.js

```diff
@@ -11,86 +11,107 @@ const getName = (parent, prop) => {
+        // it's an object and should be treated as one.
+        if (!('type' in value) && value.properties) {
+          value.type = 'object';
         }
 
-        if (items.type) {
-          array.push({
-            name: getName(parent, prop),
-            type: `[${capitalizeFirstLetter(items.type)}]`,
-            description: value.description,
-          });
+        if (value.type === 'object') {
+          array.push(flattenSchema(value, getName(parent, prop), level + 1));
         }
 
-        const newParent = parent ? `${parent}.` : '';
-        if (items.type === 'object') {
-          array.push(flattenSchema(items, oas, `${newParent}${prop}[]`, level + 1));
+        if (value.type === 'array' && value.items) {
+          let { items } = value;
+          if (items.$ref) {
+            items = findSchemaDefinition(items.$ref, oas);
+          }
+
+          // If `value` doesn't have an explicit `type` declaration, but has `properties` present,
+          // then it's an object and should be treated as one.
+          if (!('type' in items) && items.properties) {
+            items.type = 'object';
+          }
+
+          if (items.type) {
+            array.push({
+              name: getName(parent, prop),
+              type: `[${capitalizeFirstLetter(items.type)}]`,
+              description: value.description,
+            });
+          }
+
+          const newParent = parent ? `${parent}.` : '';
+          if (items.type === 'object') {
+            array.push(flattenSchema(items, `${newParent}${prop}[]`, level + 1));
+          }
+
+          return array;
         }
 
+        array.unshift({
+          name: getName(parent, prop),
+          type: capitalizeFirstLetter(value.type),
+          description: value.description,
+        });
+
         return array;
-      }
+      })
+    );
+  }
 
-      array.unshift({
-        name: getName(parent, prop),
-        type: capitalizeFirstLetter(value.type),
-        description: value.description,
+  function flattenSchema(obj, parent = '', level = 0) {
+    if (level > 2) {
+      return [];
+    }
+
+    if ('allOf' in obj) {
+      let allof = [];
+      obj.allOf.forEach(item => {
+        allof = allof.concat(flattenSchema(item));
       });
 
-      return array;
-    })
-  );
-}
+      return allof;
+    } else if ('oneOf' in obj) {
+      // We can't merge flatten objects in a `oneOf` representation into a single structure because that wouldn't
+      // validate against one of the objects, so let's just pick the first one present and flatten only that one.
+      return flattenSchema(obj.oneOf.shift());
+    } else if ('anyOf' in obj) {
+      // We can't merge flatten objects in an `anyOf` representation into a single structure because that wouldn't
+      // validate against one of the objects, so let's just pick the first one present and flatten only that one.
+      return flattenSchema(obj.anyOf.shift());
```

Comment:
Can we add `TODO`s here so that we track support for the future?

Reply:
Added a better comment with some links instead.